akka-docs/src/test/scala/docs/stream/operators/flow/StatefulMap.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package docs.stream.operators.flow
+import akka.actor.ActorSystem
+import akka.stream.scaladsl.Source
+
+object StatefulMap {
+
+  implicit val actorSystem: ActorSystem = ???
+
+  def indexed(): Unit = {
+    //#zipWithIndex
+    Source(List("A", "B", "C", "D"))
+      .statefulMap(() => 0L)((index, elem) => (index + 1, (elem, index)), _ => None)
+      .runForeach(println)
+    //prints
+    //(A,0)
+    //(B,1)
+    //(C,2)
+    //(D,3)
+    //#zipWithIndex
+  }
+
+  def bufferUntilChanged(): Unit = {
+    //#bufferUntilChanged
+    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
+      .statefulMap(() => List.empty[String])(
+        (buffer, element) =>
+          buffer match {
+            case head :: _ if head != element => (element :: Nil, buffer)
+            case _                            => (element :: buffer, Nil)
+          },
+        buffer => Some(buffer))
+      .filter(_.nonEmpty)
+      .runForeach(println)
+    //prints
+    //List(A)
+    //List(B, B)
+    //List(C, C, C)
+    //List(D)
+    //#bufferUntilChanged
+  }
+
+  def distinctUntilChanged(): Unit = {
+    //#distinctUntilChanged
+    Source("A" :: "B" :: "B" :: "C" :: "C" :: "C" :: "D" :: Nil)
+      .statefulMap(() => Option.empty[String])(
+        (lastElement, elem) =>
+          lastElement match {
+            case Some(head) if head == elem => (Some(elem), None)
+            case _                          => (Some(elem), Some(elem))
+          },
+        _ => None)
+      .collect { case Some(elem) => elem }
```

Comment:
funny, this would have been pretty good with `statefulMapConcat` since `Nil` is like `None` and `List(elem)` is like `Some(elem)`